### ISSUE TITLE
Compiler bug: NULL name when writting standard_metadata_t fields from control block instance

### ISSUE THREAD

**Issue by gkrls** [p4lang/p4c#4081]
When writing a `standard_metadata_t` field in a control block that is used by the main Ingress/Egress control I get a `NULL name` error for that field. For instance, the attached program will raise the following:

```
In file: .../_deps/p4c-src/backends/bmv2/common/expression.cpp:318
Compiler Bug: _deps/p4c-build/p4include/v1model.p4(102): NULL name: mcast_grp
    bit<16> mcast_grp;
            ^^^^^^^^^
```

Both direct type invocation and explicit instantiation of the control will raise that error. I've tested this on: 
- v1.2.3.0 - 1576090, deb package on Ubuntu 20.04, and
- v1.2.4.0 - f3667ee19, on macos 12.6
[null_field_standard_metadata.txt](https://github.com/p4lang/p4c/files/12153080/null_field_standard_metadata.txt)

**Comment by mihaibudiu** [comment]
What command-line have you used?

**Comment by mihaibudiu** [comment]
`./p4c-bm2-ss program.p4` will reproduce the issue.
This seems to happen because after inlining there is an unused instantiation of a control left which is unreachable.
This is how the code looks before the backend:
```
control __test_control(inout headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    apply {
        meta.test = 8w42;
        hdr.ethernet.etherType = standard_metadata.mcast_grp;
        standard_metadata.mcast_grp = 16w42;
    }
}

control MyIngress(inout headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    @name("MyIngress.__test_control") __test_control() __test_control_inst_0;
    apply {
        meta.test = 8w42;
        hdr.ethernet.etherType = standard_metadata.mcast_grp;
        standard_metadata.mcast_grp = 16w42;
    }
}
```

`__test_control` has been inlined, but there is still an instantiation that is using it. That instantiation should have been removed by some dead code elimination pass.

**Comment by gkrls** [comment]
Sorry i forgot the command. I indeed used `p4c-bm2-ss program.p4` as you mentioned. By the way, how did you get the output before the backend? I guess its with `--top4`, but i haven't managed to make it work.

**Comment by jafingerhut** [comment]
Here are some examples that I have successfully used to get `--top4` option to work.  I think it might require the `--dump <directory_name>` option to also be given:

+ https://github.com/jafingerhut/p4-guide/blob/master/bin/p4c-dump-few-passes.sh#L27-L44
+ https://github.com/jafingerhut/p4-guide/blob/master/bin/p4c-dump-many-passes.sh

**Comment by gkrls** [comment]
Ah great, adding `--dump` worked. Thanks Andy

**Comment by mihaibudiu** [comment]
The reason that the control is not removed is that its name starts with a double underscore, and the compiler reserves these names for internal use. If you rename your control the bug goes away.

**Comment by mihaibudiu** [comment]
Here is the relevant code, for reference:
```
const IR::Node *RemoveUnusedDeclarations::process(const IR::IDeclaration *decl) {
    LOG3("Visiting " << decl);
    if (decl->getName().name == IR::ParserState::verify && getParent<IR::P4Program>())
        return decl->getNode();
    if (decl->getName().name.startsWith("__"))
        // Internal identifiers, e.g., __v1model_version
        return decl->getNode();
    if (refMap->isUsed(getOriginal<IR::IDeclaration>())) return decl->getNode();
    LOG3("Removing " << getOriginal());
    prune();  // no need to go deeper
    return nullptr;
}
```

**Comment by fruffy** [comment]
Still, the compiler should throw an error and not a compiler bug, no? Is there an easy way to throw an error here?

**Comment by mihaibudiu** [comment]
It's a bit difficult to anticipate all illegal user behaviors. Moreover, it's not easy to give a good explanation to the user about what the error is.

**Comment by fruffy** [comment]
Considering code like 

```
    if (decl->getName().name.startsWith("__"))
        // Internal identifiers, e.g., __v1model_version
        return decl->getNode();
```

It might make sense to include a compiler pass that checks whether this declaration is part of the legal internal identifiers. If not, throw an error? Or potentially make it a separate token with its own IR class. I do not know of the usage of these internal identifiers.

**Comment by mihaibudiu** [comment]
The spec reserves these for internal use for libraries and backends.
There is no such list.

**Comment by fruffy** [comment]
> The spec reserves these for internal use for libraries and backends.

Yes, so these are either part of the spec or compiler front end or added by back ends. It is possible to just maintain a list of these, right? For back ends, it might be possible to extend that list.

**Comment by mihaibudiu** [comment]
but the pass which removes declarations is in the front-end.
There is no API by which the backend pass information to the front-end.
Frankly, I think there are much more important problems to worry about than this one. It would be nice to have, but I'd rather invest the time fixing a bug or implementing a new feature.

**Comment by fruffy** [comment]
The API is build-system-specific, e.g., CMake can definitely do this.  So I'd say it is possible to implement this very early. I would leave this as an issue we could implement some day.

**Comment by mihaibudiu** [comment]
It was supposed to be in the spec, but I can't find it. Perhaps we never wrote it down.

**Comment by jafingerhut** [comment]
Perhaps you are thinking about this issue, and any PRs that might have been created to address it?  https://github.com/p4lang/p4-spec/issues/929
